src/classes/ADDR_Contact_TDTM.cls

```diff
@@ -195,40 +197,67 @@ public class ADDR_Contact_TDTM extends TDTM_Runnable {
                ) {
                 // note that we detect whether the address change is coming from the address trigger,
                 // in which case we don't want to create a new address.
-                if (ADDR_Addresses_TDTM.hasRunAddrTrigger) {
+                if (TDTM_ProcessControl.isRecursive(TDTM_ProcessControl.flag.ADDR_hasRunTrigger)) {
                 	//UTIL_Debug.debug('DJH**** stop Addr Trigger recursion in ADDR_Contact_TDTM');
                     continue;
                 }
+                // If the Before Insert/Update trigger on the Contact has already executed and taken action
+                // and this is an After Insert/Update trigger, simply reset the flag and exit out.
+                if (TDTM_ProcessControl.isRecursive(TDTM_ProcessControl.flag.ADDR_hasRunContactOnBefore) &&
```

Comment:
@xinnanlisf This (`ADDR_hasRunContactOnBefore`) is the new flag/logic related to handling the Contact part of the address trigger. This is set/referenced in a few different places in the code base to ensure that recursion does not occur.